python/tests/test_conversions.py

```diff
@@ -0,0 +1,288 @@
+        (DecimalType(31, 26), b"\x0f\x94\xec\xad\xa5C<]\xdePf\xd6N", Decimal("12345.12345678912345678912345678")),
+    ],
+)
+def test_round_trip_conversion_large_decimals(primitive_type, b, result):
+    """Test round trip conversions of calling `conversions.from_bytes` and then `conversions.to_bytes` on the result"""
+    value_from_bytes = conversions.from_bytes(primitive_type, b)
+    assert value_from_bytes == result
+
+    bytes_from_value = conversions.to_bytes(primitive_type, value_from_bytes)
+    assert bytes_from_value == b
+
+
+@pytest.mark.parametrize(
+    "primitive_type, b, result",
+    [
+        (FloatType(), b"\x19\x04\x9e?", 1.2345),
+        (DoubleType(), b"\x8d\x97n\x12\x83\xc0\xf3?", 1.2345),
+    ],
+)
+def test_round_trip_conversion_approximation(primitive_type, b, result):
+    """Test approximate round trip conversions of calling `conversions.from_bytes` and then `conversions.to_bytes` on the result"""
+    value_from_bytes = conversions.from_bytes(primitive_type, b)
+    assert value_from_bytes == pytest.approx(result)
+
+    bytes_from_value = conversions.to_bytes(primitive_type, value_from_bytes)
+    assert bytes_from_value == b
+
+
+def test_raise_on_unregistered_type():
+    """Test raising when a conversion is attempted for a type that has no registered method"""
+
+    class FooUnknownType:
+        def __repr__(self):
+            return "FooUnknownType()"
+
+    with pytest.raises(TypeError) as exc_info:
+        conversions.from_partition_value_to_py(FooUnknownType(), "foo")
+    assert (f"Cannot convert partition string to python built-in, type FooUnknownType() not supported: 'foo'") in str(
+        exc_info.value
+    )
+
+    with pytest.raises(TypeError) as exc_info:
+        conversions.to_bytes(FooUnknownType(), "foo")
+    assert ("Cannot serialize value, type FooUnknownType() not supported: 'foo'") in str(exc_info.value)
+
+    with pytest.raises(TypeError) as exc_info:
+        conversions.from_bytes(FooUnknownType(), b"foo")
+    assert ("Cannot deserialize bytes, type FooUnknownType() not supported: b'foo'") in str(exc_info.value)
+
+
+@pytest.mark.parametrize(
+    "primitive_type, value, expected_error_message",
+    [
+        (DecimalType(7, 3), Decimal("123.4567"), "Cannot serialize value, scale of value does not match type decimal(7, 3): 4"),
+        (
+            DecimalType(18, 8),
+            Decimal("123456789.123456789"),
+            "Cannot serialize value, scale of value does not match type decimal(18, 8): 9",
+        ),
+        (
+            DecimalType(36, 34),
+            Decimal("1.23456789123456789123456789123456789"),
+            "Cannot serialize value, scale of value does not match type decimal(36, 34): 35",
+        ),
+        (
+            DecimalType(7, 2),
+            Decimal("1234.56"),
+            "Cannot serialize value, precision of value does not match type decimal(7, 2): 6",
```

Comment:
To follow-up on my question on the implementation. How can a user specifying this value?  It seems like it should only be an error if precision is exceeded (as you test for below).

Reply:
Replied [here](https://github.com/apache/iceberg/pull/4263#discussion_r822026954) and fixed!